### ISSUE TITLE
Ribbon integration keeps reference to "old" ApplicationContexts even when using @DirtiesContext

### ISSUE THREAD

**Issue by michele-mancioppi** [spring-cloud/spring-cloud-netflix#2147]
The `ApplicationContext` used by `HttpClientRibbonCommandFactory` of `RibbonRoutingFilter` does not seem to be refreshed by the `@DirtiesContext` annotation on SpringBoot tests.

This causes tests using `@DirtiesContext` to fail because all tests with a "dirty" context are trying to used an inactive and closed `ApplicationContext`, which means that this exception is thrown: https://github.com/spring-projects/spring-framework/blob/v4.3.4.RELEASE/spring-context/src/main/java/org/springframework/context/support/AbstractApplicationContext.java#L1065

Now, it does not exactly sound like a cause that `@DirtiesContext` should be able to handle out of the box (though it'd be nice). What would be the best way to refresh the `HttpClientRibbonCommandFactory` of a `RibbonRoutingFilter` across test runs?

**Comment by spencergibb** [comment]
Can you tell me the symptoms you are having when you run the test? We have many tests using zuul and ribbon that work ok.

**Comment by michele-mancioppi** [comment]
The symptom is that the line https://github.com/spring-projects/spring-framework/blob/v4.3.4.RELEASE/spring-context/src/main/java/org/springframework/context/support/AbstractApplicationContext.java#L1065 (but put here also the latest & greatest spring release) is invoked, an  IllegalStateException with message "ApplicationContext@... has been closed already" is thrown and the test fails :-)

Do your tests with Zuul and Rubbon require the refreshing of the ApplicationContext? (We need to change the environment between tests.) Do you use @DirtiesContext in them? Can you share an example of such test?

**Comment by spencergibb** [comment]
See https://github.com/spring-cloud/spring-cloud-netflix/blob/master/spring-cloud-netflix-core/src/test/java/org/springframework/cloud/netflix/zuul/ZuulProxyApplicationTests.java

It might be easier to submit a small application with a failing test.

**Comment by PatrikSteuer** [comment]
You can reproduce the issue with the mentioned class, just change the classmode of the DirtiesContext 
`@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)`
and add following depdenency to your pom 
 ```
       <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-undertow</artifactId>
        </dependency>
```

**Comment by spencergibb** [comment]
I'd like to hear from @michele-mancioppi. Your suggestiong @PatrikSteuer seems a bit artificial.

**Comment by michele-mancioppi** [comment]
Hi @spencergibb, @PatrikSteuer is a team-mate of mine and we were together spending a lot of "quality" time debugging this ;-)

What is it that you find artificial in @PatrikSteuer 's suggestion?

**Comment by spencergibb** [comment]
So `@DirtiesContext` works, `@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)` doesn't work?

**Comment by michele-mancioppi** [comment]
No. As reported in the description: `ApplicationContext` used by `HttpClientRibbonCommandFactory` of `RibbonRoutingFilter` does not seem to be refreshed by the `@DirtiesContext` annotation on SpringBoot tests.

That is, no matter what the `DirtiesContext.ClassMode` is, the `ApplicationContext` used by `HttpClientRibbonCommandFactory`is not affected, and that causes *subsequent* tests to fail. In this context, *subsequent* of course depends on which `DirtiesContext.ClassMode` we use.

**Comment by spencergibb** [comment]
> You can reproduce the issue with the mentioned class,

Which class is that, `ZuulProxyApplicationTests.java`?

**Comment by PatrikSteuer** [comment]
Exactly. I thought it would be the most easiest way to reproduce the problem by referencing to a class you allready know.
